software/controller/lib/core/alarm.cpp

```diff
@@ -0,0 +1,102 @@
+#include <stdint.h>
+#include "vars.h"
+#include "debug.h"
+#include "hal.h"
+#include "alarm.h"
+#include "comms.h"
+#include "commfail_alarm.h"
+
+// MAX_CONTROLLER_ALARMS 
+#define MAX_CONTROLLER_ALARMS 2
+
+// structure definition for MAX_CONTROLLER_ALARMS 
+alarm_cb_t alarms[MAX_CONTROLLER_ALARMS];
+
+// this value is reset to 0 after MAX_CONTROLLER_ALARMS 
+static uint16_t alarm_check_counter;
+
+// There should be a packet from GUI for every 30ms. Assuming something went
+// wrong, we wait for 50ms before raising a communication failure alarm
+static constexpr Duration COMMUNICATION_TIMEOUT = milliseconds(10000);
+static DebugUInt32 dbg_alarm_check("alarm_check",
+                                      "alarm_check",
+                                      0);
+uint32_t alarm_check;
+
+extern Time last_rx;
+extern Time comms_get_lastrx_time(void);
+
+static bool is_time_to_raise_alarm() {
+    return Hal.now() - comms_get_lastrx_time() > COMMUNICATION_TIMEOUT;
+}
+
+void alarm_init(void)
+{
+    // for COMMUNICATION_CHECK_ALARM 1
+    alarms[COMM_CHECK_ALARM].timestamp = Hal.now();
+    alarms[COMM_CHECK_ALARM].callthis = commfail_alarm_cb;
+    alarms[COMM_CHECK_ALARM].priority = ALARM_HIGH_PRIORITY;
+    alarms[COMM_CHECK_ALARM].triggered = 0;
+    alarms[COMM_CHECK_ALARM].alarm_pause = 0;// 0 not pausable, 1 pausable
+    alarms[COMM_CHECK_ALARM].audio_pause = 0;// 0 not pausable, 1 pausable
+    alarms[COMM_CHECK_ALARM].trigger_count = 0;
+    // for UNDEFINED_ALARM 2
+    alarms[UNDEFINED_ALARM].timestamp = Hal.now();
+    alarms[UNDEFINED_ALARM].callthis = NULL;
+    alarms[UNDEFINED_ALARM].priority = 0;
+    alarms[UNDEFINED_ALARM].triggered = 0;
+    alarms[UNDEFINED_ALARM].alarm_pause = 0;
+    alarms[UNDEFINED_ALARM].audio_pause = 0;
+    alarms[UNDEFINED_ALARM].trigger_count = 0;
+    alarm_check_counter = 0;
+    alarm_check = dbg_alarm_check.Get();
+    debug.Print("alarm_init");
+}
+
+void alarm_handler(void)
+{
+    static uint32_t count;
+
+    alarm_check = dbg_alarm_check.Get();
+    if (alarm_check_counter < MAX_CONTROLLER_ALARMS) {
```

Comment:
The logic below seems to indicated that an alarm will be triggered if `is_time_to_raise_alarm` is true OR if `dbg_alarm_check.Get()` has returned true, and in either case, the `alarm_check_counter` will be triggered? But if it was the first case and alarm was triggered due to comms timeout, then it was not an alarm check, so then this variable is not appropriately named. Or am I misunderstanding something?

Reply:
this is not needed. dbg_alarm_check.Get()  is for my debug purpose. I removed this in the present code.